# Project: Database

**Instructions for Students:**

Please carefully follow these steps to complete and submit your project:

1. **Make a copy of the Project**: Please make a copy of this project either to your own Google Drive or download locally. Work on the copy of the project. The master project is **Read-Only**, meaning you can edit, but it will not be saved when you close the master project. To avoid total loss of your work, remember to make a copy.

2. **Completing the Project**: You are required to work on and complete all tasks in the provided project. Be disciplined and ensure that you thoroughly engage with each task.
   
3. **Creating a Google Drive Folder**: Each of you must create a new folder on your Google Drive. This will be the repository for all your completed project files, aiding you in keeping your work organized and accessible.
   
4. **Uploading Completed Project**: Upon completion of your project, make sure to upload all necessary files, involving codes, reports, and related documents into the created Google Drive folder. Save this link in the 'Student Identity' section and also provide it as the last parameter in the `submit` function that has been provided.
   
5. **Sharing Folder Link**: You're required to share the link to your project Google Drive folder. This is crucial for the submission and evaluation of your project.
   
6. **Setting Permission to Public**: Please make sure your Google Drive folder is set to public. This allows your instructor to access your solutions and assess your work correctly.

Adhering to these procedures will facilitate a smooth project evaluation process for you and the reviewers.

## Project Description

In this project, you'll work with the Kaggle E-commerce dataset. Your tasks are to import this dataset into a SQLite database, explore it using SQL to identify key attributes and patterns, and perform basic data cleaning procedures. This project is designed to improve your skills in database management, SQL usage, and data cleaning.

## Grading Criteria

There are 13 criterias in this project, all have the same weight divided into two: 8 criterias for SQL and 5 criterias for Elasticssearch. Each task will give you either 100 point if you are correct and 0 if you are wrong. The final score for the project will the the average of all 13 tasks.

* Criteria 1, Task I.3: This task will assess your ability to create an SQL database, in this case using sqlite. To ensure correct data is imported, this will check that database table have the correct amount of columns and the correct column names. This will ensure a uniform dataset for all students.

* Criteria 2, Task I.3: Same as Criteria 1, instead of columns, this will check that the total rows of data in the table is correct. This will ensure a uniform dataset for all students.

* Criteria 3, Task I.3.4 query 1: The task will assess your ability to understand and perform basic dataset check, that one row in the database does not guarantee uniqueness for individual columns.

* Criteria 4, Task I.3.4 query 1: Same as Criteria 3.

* Criteria 5, Task I.3.4 query 2: This task will assess your ability to create a SQL query that will get the data, perform arithmetic operations on the data, sort and limit the data returned.

* Criteria 6, Task I.3.4 query 3: Same as Criteria 5.

* Criteria 7, Task I.3.4 query 4: This task will assess your ability to create a SQL query that will get the data, perform arithmetic operations on the data, sort, limit the data and structure the result in a concise format where the all the data are retrieved and presented in a format that is easy to understand can be further processed through programming.

* Criteria 8, Task I.3.4 query 5: Same as Criteria 7.

* Criteria 9, Task II.3: This task will assess your ability to create an Elasticsearch Index. To ensure correct data is imported, this will check that the index have the correct amount of columns the the correct column names. This will ensure a uniform dataset for all students.

* Criteria 10, Task II.3: Same as Criteria 9, instead of columns, this will check that the total rows of data in the table is correct. This will ensure a uniform dataset for all students.

* Criteria 11, Task II.4 query 1: This task will assess your ability to perform a basic elastic search and return the result.

* Criteria 12, Task II.4 query 2: This task will assess your ability to perform an elastic search, perform arithmetic operations on the data and return the result.

* Criteria 13, Task II.4 query 3: This task will assess your ability to perform an elastic search, perform arithmetic operations on the data, sort and limit the data returned.


## Student Identity

In [17]:
# @title #### Student Identity
student_id = "REAPJWFQ" # @param {type:"string"}
name = "Fathur Imam Mujaddid" # @param {type:"string"}
drive_link = "https://drive.google.com/drive/folders/1Ld7aR8s0C7X36Mala-WyeYThXd7tJX3K?usp=sharing"  # @param {type:"string"}

assignment_id = "00_database_project"

# Import grader package
!pip install rggrader
from rggrader import submit, submit_image

Rmember to make a copy of this notebook in your Google Drive and work in your own copy.

## I. Loading and Querying with SQL

In this task, you will use Kaggle E-commerce dataset. You need to download it from the link here: [Kaggle E-commerce data](https://www.kaggle.com/datasets/carrie1/ecommerce-data?select=data.csv).

After you download the dataset, you will then import the dataset into SQL format, using SQLite.

### I.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [18]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below
import sqlite3
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### I.2. Data Loading and SQL CREATE and INSERT to database table

Now you can create the database and tables. Follow the same column naming as the dataset.

Next insert all the data from the dataset into the table in the database.

In [19]:
# Write your data loading to SQL database here
dataset = pd.read_csv('/content/drive/MyDrive/AI Engineer Bootcamp/Project/5. Database/data.csv', encoding='ISO-8859-1')
# dataset = pd.read_csv('data.csv', encoding='ISO-8859-1')
dataset.head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [ ]:
# Buat atau koneksi ke database SQLite
conn = sqlite3.connect('project_5.db')

# Unggah data dari DataFrame ke SQLite (Pandas akan membuat tabel otomatis)
dataset.to_sql('dataset_p5', conn, if_exists='replace', index=False)

# Commit perubahan dan tutup koneksi
conn.commit()
conn.close()

In [ ]:
import sqlite3

# Connect to SQLite database
conn = sqlite3.connect('project_5.db')
cur = conn.cursor()

# Get list of tables
cur.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = cur.fetchall()

print(f"Tables: {tables}")

# Close the connection
conn.close()


Tables: [('dataset_p5',)]


### I.3. Verifying the database

Before we go to the next section, let's check our database first. You need to change the db_name and table_name with your database name and table name, afterwards you can just run the code block. The result should be:

- Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
- Number of rows: 541909

In [20]:
import sqlite3

# Connect to SQLite database
conn = sqlite3.connect('project_5.db') #change db_name to your database name.
cur = conn.cursor()

# Get column names
cur.execute("PRAGMA table_info(dataset_p5)") #change table_name to your table name
sql_columns = [column[1] for column in cur.fetchall()] #the variable sql_columns is used for submission, do not change the variable name
print(f"Columns: {sql_columns}")

# Get total number of rows
cur.execute("SELECT COUNT(*) FROM dataset_p5") #change table_name to your table name
sql_num_rows = cur.fetchone()[0] #the variable sql_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {sql_num_rows}")

# Close the connection
conn.close()

Columns: ['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']
Number of rows: 541909


### I.3.4. Running SQL Queries

Once we have the database in SQL format, let's do some data exploration and analysis.

You are part of the engineering team at a global e-commerce company. Your company has a vast catalogue of products and serves customers around the globe. As the company continues to grow, the management decides it's time to expand the business even more.

As an essential part of this expansion, your team is tasked with diving into your company's rich e-commerce dataset to extract crucial insights that will guide the expansion strategy.

1. **Total number of unique products and unique customers:**

   Your first task is to gauge the breadth of your operations. By determining the total number of unique products you sell and the total number of unique customers you serve, you can assess the scale and diversity of your business.

2. **Total revenue for each product:**

   Next, you're going to identify the star performers in your product portfolio. Calculating the total revenue for each product will reveal which items are the biggest revenue drivers. These insights can help guide decisions about product focus and marketing efforts. Limit the result to the top 5 product with the biggest revenue.

3. **Most profitable countries (by total sales):**

   To pinpoint the most fruitful geographic areas for your business, you'll identify which countries generate the most profit. This data will help you understand where your operations are already strong and where there might be potential for regional expansion. Limit the result to the top 5 countries with the biggest profit.

4. **Top-selling products (by total sales) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by sales for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

4. **Top-selling products (by quantity) for each country from number 3:**

   Product popularity may vary across different regions. Determining the top-selling products by quantity for each country will provide a clear picture of regional preferences, informing decisions about product distribution in different markets. Limit the result to the top 5 products for each countries.

By answering these queries, your team will provide valuable data-driven insights that will directly contribute to the strategic decisions about where and how your business expands. As part of the engineering team, you're not just maintaining the technical infrastructure - you're shaping the future of the company.

> Hint: The Product name is stored in column Descrioption

In [ ]:
# Write your SQL Query here

#use the following variable name to store the result from your SQL Query
sql_num_products = 0
sql_num_customers = 0
sql_top_product_revenues = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_countries = [] # list of tuples, where each tupple contains the product name and total revenue/total sales, see example
sql_top_selling_products_by_country_by_sales = {} #dictionary where key is country name, value is a list of top-selling products by sales, see example below
sql_top_selling_products_by_country_by_quantity = {} #dictionary where key is country name, value is a list of top-selling products by quantity, see example below


In [ ]:
conn = sqlite3.connect('project_5.db') #change db_name to your database name.
cur = conn.cursor()
# 1. Find the total number of unique products and unique customers
# Example: sql_num_products = 1559
# Example: sql_num_customers = 1675
x = pd.read_sql_query("SELECT COUNT(DISTINCT StockCode) AS count_StockCode FROM dataset_p5", conn)
y = pd.read_sql_query("SELECT COUNT(DISTINCT CustomerID) AS count_costumerid FROM dataset_p5", conn)
conn.commit()
conn.close()
print(x)
print(y)

   count_StockCode
0             4070
   count_costumerid
0              4372


In [ ]:
sql_num_products = 4070
sql_num_customers = 4372


In [ ]:
# 2. Calculate the total revenue for each product, limited to top 5
# Example: sql_top_product_revenues = [('PRODUCT 1', 149385.12), ('PRODUCT 2', 54895.53), ('PRODUCT 3', 44545.55), ('PRODUCT 4', 38000.00), ('PRODUCT 5', 36000.00)]

import sqlite3

# Connect to SQLite database
conn = sqlite3.connect('project_5.db')
cur = conn.cursor()

# Query to calculate total revenue for each product and limit to top 5
cur.execute('''
    SELECT Description, SUM(Quantity * UnitPrice) AS TotalRevenue
    FROM dataset_p5
    GROUP BY Description
    ORDER BY TotalRevenue DESC
    LIMIT 5;
''')

# Use fetchall() to retrieve all rows from the result set
sql_top_product_revenues = cur.fetchall()

# Print the top 5 products by revenue
print(sql_top_product_revenues)

# Close the connection
conn.close()



[('DOTCOM POSTAGE', 206245.48000000019), ('REGENCY CAKESTAND 3 TIER', 164762.19), ('WHITE HANGING HEART T-LIGHT HOLDER', 99668.46999999849), ('PARTY BUNTING', 98302.98000000048), ('JUMBO BAG RED RETROSPOT', 92356.03000000204)]


In [ ]:
# 3. Identify the most profitable countries (by total sales), limited to top 5
# Example: sql_top_countries = [('COUNTRY 1', 600000.00), ('COUNTRY 2', 334857.45), ('COUNTRY 3', 245879.00), ('COUNTRY 4', 180000.00), ('COUNTRY 5', 164389.45)]

# Connect to SQLite database
conn = sqlite3.connect('project_5.db')
cur = conn.cursor()

# Query to calculate total revenue for each country and limit to top 5
cur.execute('''
    SELECT Country, SUM(Quantity * UnitPrice) AS TotalRevenueByCountry
    FROM dataset_p5
    GROUP BY Country
    ORDER BY TotalRevenueByCountry DESC
    LIMIT 5;
''')

# Use fetchall() to retrieve all rows from the result set
sql_top_countries = cur.fetchall()

# Print the top 5 products by revenue
print(sql_top_countries)

# Close the connection
conn.close()


[('United Kingdom', 8187806.364001113), ('Netherlands', 284661.54000000015), ('EIRE', 263276.81999999826), ('Germany', 221698.20999999862), ('France', 197403.90000000002)]


In [21]:
# 4. Find the top-selling products (by total sales) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_by_sales = {
#    'COUNTRY 1': [('PRODUCT 1', 200000.00), ('PRODUCT 2', 100000.00), ('PRODUCT 3', 95348.00), ('PRODUCT 4', 90000.43), ('PRODUCT 5', 80000.00)],
#    'COUNTRY 2': [('PRODUCT 1', 9000.00), ('PRODUCT 2', 7345.00), ('PRODUCT 3', 6934.23), ('PRODUCT 4', 6800.23), ('PRODUCT 5', 4000.00)],
#    ... to simplify, country 3 and 4 are not shown
#    'COUNTRY 5': [('PRODUCT 1', 7442.84), ('PRODUCT 2', 5175.00), ('PRODUCT 3', 3098.0), ('PRODUCT 4', 3014.34), ('PRODUCT 5', 2857.83)],



In [22]:
import sqlite3

# Connect to SQLite database
conn = sqlite3.connect('project_5.db')
cur = conn.cursor()

# Query to find the top-selling products for each country, limited to top 5 products per country
cur.execute('''
    WITH CountrySales AS (
        SELECT Country, SUM(Quantity * UnitPrice) AS TotalRevenueByCountry
        FROM dataset_p5
        GROUP BY Country
        ORDER BY TotalRevenueByCountry DESC
        LIMIT 5
    ),
    ProductSales AS (
        SELECT Country, Description, SUM(Quantity * UnitPrice) AS TotalRevenueByProduct
        FROM dataset_p5
        GROUP BY Country, Description
    )
    SELECT cs.Country, ps.Description, ps.TotalRevenueByProduct
    FROM ProductSales ps
    JOIN CountrySales cs ON ps.Country = cs.Country
    ORDER BY cs.TotalRevenueByCountry DESC, ps.TotalRevenueByProduct DESC;
''')

# Fetch all results
results = cur.fetchall()

# Process results into the desired format
sql_top_selling_products_by_country_by_sales = {}
for country, product, revenue in results:
    if country not in sql_top_selling_products_by_country_by_sales:
        sql_top_selling_products_by_country_by_sales[country] = []
    sql_top_selling_products_by_country_by_sales[country].append((product, revenue))

# Limit to top 5 products for each country
for country in sql_top_selling_products_by_country_by_sales:
    sql_top_selling_products_by_country_by_sales[country] = sql_top_selling_products_by_country_by_sales[country][:5]



# Close the connection
conn.close()


In [23]:

sql_top_selling_products_by_country_by_sales

{'United Kingdom': [('DOTCOM POSTAGE', 206245.48000000019),
  ('REGENCY CAKESTAND 3 TIER', 134405.93999999992),
  ('WHITE HANGING HEART T-LIGHT HOLDER', 93953.0699999987),
  ('PARTY BUNTING', 92501.73000000033),
  ('JUMBO BAG RED RETROSPOT', 84516.44000000166)],
 'Netherlands': [('RABBIT NIGHT LIGHT', 9568.48),
  ('ROUND SNACK BOXES SET OF4 WOODLAND ', 7991.400000000001),
  ('SPACEBOY LUNCH BOX ', 7485.5999999999985),
  ('DOLLY GIRL LUNCH BOX', 6828.599999999999),
  ('ROUND SNACK BOXES SET OF 4 FRUITS ', 4039.2000000000007)],
 'EIRE': [('REGENCY CAKESTAND 3 TIER', 7442.849999999997),
  ('CARRIAGE', 5175.0),
  ('JAM MAKING SET WITH JARS', 3089.0),
  ('3 TIER CAKE TIN RED AND CREAM', 3041.5499999999997),
  ('WHITE HANGING HEART T-LIGHT HOLDER', 2857.8)],
 'Germany': [('POSTAGE', 20821.0),
  ('REGENCY CAKESTAND 3 TIER', 8257.349999999997),
  ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3554.6999999999985),
  ('ROUND SNACK BOXES SET OF 4 FRUITS ', 1949.9500000000016),
  ('SPACEBOY LUNCH BOX ', 

In [24]:
# 5. Find the top-selling products (by quantity) for each country, limited to top 5
# Example: sql_top_selling_products_by_country_bNTRY 2 PRODUCT 1', 9000), ('COUNTRY 2 PRODUCT 2', 7345), ('COUNTRY 2 PRODUCT 3', 6934), ('COUNTRY 2 PRODUCT 4', 6800), ('COUNTRY 2 PRODUCT 5', 4000)],
#    ... to simply_quantity = {
#    'COUNTRY 1': [('COUNTRY 1 PRODUCT 1', 22343), ('COUNTRY 1 PRODUCT 2', 12345), ('COUNTRY 1 PRODUCT 3', 9534), ('COUNTRY 1 PRODUCT 4', 9000), ('COUNTRY 1 PRODUCT 5', 8000)],
#    'COUNTRY 2': [('COUify, country 3 and 4 are not shown
#    'COUNTRY 5': [('COUNTRY 5 PRODUCT 1', 48345), ('COUNTRY 5 PRODUCT 2', 43965), ('COUNTRY 5 PRODUCT 3', 43556), ('COUNTRY 5 PRODUCT 4', 34567), ('COUNTRY 5 PRODUCT 5', 33436)],


In [25]:
import sqlite3

# Connect to SQLite database
conn = sqlite3.connect('project_5.db')
cur = conn.cursor()

# Query to find the top-selling products for each country, limited to top 5 products per country
cur.execute('''
    WITH CountrySales AS (
        SELECT Country, SUM(Quantity) AS TotalQuantityByCountry
        FROM dataset_p5
        GROUP BY Country
        ORDER BY TotalQuantityByCountry DESC
        LIMIT 5
    ),
    ProductSales AS (
        SELECT Country, Description, SUM(Quantity) AS TotalQuantityByProduct
        FROM dataset_p5
        GROUP BY Country, Description
    )
    SELECT cs.Country, ps.Description, ps.TotalQuantityByProduct
    FROM ProductSales ps
    JOIN CountrySales cs ON ps.Country = cs.Country
    ORDER BY cs.TotalQuantityByCountry DESC, ps.TotalQuantityByProduct DESC;
''')

# Fetch all results
results = cur.fetchall()

# Process results into the desired format
sql_top_selling_products_by_country_by_quantity = {}
for country, product, quantity in results:
    if country not in sql_top_selling_products_by_country_by_quantity:
        sql_top_selling_products_by_country_by_quantity[country] = []
    sql_top_selling_products_by_country_by_quantity[country].append((product, quantity))

# Limit to top 5 products for each country
for country in sql_top_selling_products_by_country_by_quantity:
    sql_top_selling_products_by_country_by_quantity[country] = sql_top_selling_products_by_country_by_quantity[country][:5]


# Close the connection
conn.close()


In [26]:
sql_top_selling_products_by_country_by_quantity

{'United Kingdom': [('WORLD WAR 2 GLIDERS ASSTD DESIGNS', 48326),
  ('JUMBO BAG RED RETROSPOT', 43167),
  ('POPCORN HOLDER', 34365),
  ('ASSORTED COLOUR BIRD ORNAMENT', 33679),
  ('WHITE HANGING HEART T-LIGHT HOLDER', 33193)],
 'Netherlands': [('RABBIT NIGHT LIGHT', 4801),
  ('SPACEBOY LUNCH BOX ', 4528),
  ('DOLLY GIRL LUNCH BOX', 4132),
  ('PACK OF 72 RETROSPOT CAKE CASES', 4128),
  ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3132)],
 'EIRE': [('PACK OF 72 RETROSPOT CAKE CASES', 1728),
  ('60 TEATIME FAIRY CAKE CASES', 1536),
  ('VINTAGE SNAP CARDS', 1492),
  ('ASSORTED INCENSE PACK', 1440),
  ('PACK OF 60 PINK PAISLEY CAKE CASES', 1296)],
 'Germany': [('ROUND SNACK BOXES SET OF4 WOODLAND ', 1218),
  ('ASSORTED COLOURS SILK FAN', 1164),
  ('POSTAGE', 1104),
  ('WOODLAND CHARLOTTE BAG', 1019),
  ('PACK OF 72 RETROSPOT CAKE CASES', 1002)],
 'France': [('RABBIT NIGHT LIGHT', 4023),
  ('MINI PAINT SET VINTAGE ', 2196),
  ('RED TOADSTOOL LED NIGHT LIGHT', 1314),
  ('SET/6 RED SPOTTY PAPER CUP

## II. Indexing and Querying with Elasticsearch

You should have a better understanding of how to use SQL, now let's leverage the power of Elasticsearch to explore and analyze the data. We'll use the same dataset.


### II.1. Package and Module Installation

First, let's pool all package and module that you'll need in the installation section below.

In [ ]:
# Write any package/module installation that you need
# pip install goes here, this helps declutter your output below
%pip install elasticsearch
from elasticsearch import Elasticsearch
import json
import pandas as pd


Note: you may need to restart the kernel to use updated packages.


### II.2. Set up Elasticsearch Index

The first step you need to do which is **preprocessing, let's keep it simple and simply drop all row that have missing values.**

Next, you need to set up an Elasticsearch index. An Elasticsearch index is like a database in a traditional relational database. An index lets you store, search, and analyze big volumes of data quickly and in near real time.

You'll also need to setup the mappings for the data types, let's do that here.

In [ ]:
# Preprocessing here
dataset = pd.read_csv('data.csv', encoding='ISO-8859-1')
new_dataset = dataset.dropna()
# Load your dataset and create an Elasticsearch index
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])
es.indices.create(index='dataset_p5_es', ignore=400)
# Hints: use the BulkAPI for faster creation of the index
#import data to elastic
def dataframe_to_dict(new_dataset):
    records = new_dataset.to_dict(orient='records')
    return records

data_dict = dataframe_to_dict(new_dataset)

def generate_actions(data):
    for record in data:
        yield {
            "_index": "dataset_p5_es",
            "_source": record
        }

from elasticsearch.helpers import bulk
#cek
success, failed = bulk(es, generate_actions(data_dict))
print(f"Berhasil mengunggah {success} dokumen, gagal {failed} dokumen")



C:\Users\fathu\AppData\Local\Temp\ipykernel_17648\2519907067.py:6: DeprecationWarning: Passing transport options in the API method is deprecated. Use 'Elasticsearch.options()' instead.
  es.indices.create(index='dataset_p5_es', ignore=400)


Berhasil mengunggah 406829 dokumen, gagal [] dokumen


In [ ]:
new_dataset.shape

(406829, 8)

In [ ]:
# from elasticsearch import Elasticsearch

# # Koneksi ke Elasticsearch
# es = Elasticsearch("http://localhost:9200")

# # Nama indeks yang ingin dihapus
# index_name = 'dataset_p5_es'

# # Cek apakah indeks ada
# if es.indices.exists(index=index_name):
#     # Hapus indeks
#     es.indices.delete(index=index_name)
#     print(f"Index '{index_name}' deleted")

#     # Membuat ulang indeks (opsional)
#     es.indices.create(index=index_name)
#     print(f"Index '{index_name}' created")
# else:
#     print(f"Index '{index_name}' does not exist")


Index 'dataset_p5_es' deleted
Index 'dataset_p5_es' created


### II.3. Verifying the index

Before we go to the next section, let's check our index first. You need to change the my_index_name with your chosen index name, afterwards you can just run the code block. The result should be:

- Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
- Number of rows: 406829

In [ ]:
from elasticsearch import Elasticsearch

# Connect to the Elasticsearch server
es = Elasticsearch([{'host': 'localhost', 'port': 9200, 'scheme': 'http'}])

index_name = 'dataset_p5_es' #change the my_index_name value

# Get index mapping (equivalent to getting column names in SQL)
mapping = es.indices.get_mapping(index=index_name)
es_columns = list(mapping[index_name]['mappings']['properties'].keys()) #the variable es_columns is used for submission, do not change the variable name
print(f"Columns: {es_columns}")

# Count documents in the index (equivalent to counting rows in SQL)
es_num_rows = es.count(index=index_name)['count'] #the variable es_num_rows is used for submission, do not change the variable name
print(f"Number of rows: {es_num_rows}")


Columns: ['Country', 'CustomerID', 'Description', 'InvoiceDate', 'InvoiceNo', 'Quantity', 'StockCode', 'UnitPrice']
Number of rows: 406829


### II.4. Running Elasticsearch queries

Once your e-commerce data is cleaned and ready, let's analyse the data to gain insights that will drive the expansion strategy, this time using Elasticsearch and we'll focus on one specific country, which is Germany.

1. **Find all transactions for a specific country**

   Your company operates globally, but you want to understand better how different countries contribute to your sales. By finding all transactions that happened in a specific country, you can get a clearer picture of your company's reach and performance in that location. In our case, we'll focus on the country Germany.


2. **Find all unique products in Germany**

   Your first task is to understand the breadth of your company's product portfolio in Germany. By determining the total number of unique products you sell, you can assess the scale and diversity of your operations.


3. **Find the top 5 most purchased products in Germany**

   Next up, you'll identify the best-selling products in your portfolio in Germany. Knowing which items are the most purchased can help guide decisions about product focus and marketing efforts. We'll limit the result to 5 most purchased products.


By executing these tasks, you will enable your team to provide data-driven insights which will directly contribute to the strategic decisions about your business expansion. As part of the data engineering team, you're not only maintaining the technical infrastructure - you're shaping the future of the company.


In [ ]:
# Write your Elasticsearch Query here

#use the following variable name to store the result from your Elasticsearch Query
es_total_unique_products = 0
es_transactions_in_germany = 0
es_top_products = {} #dictionary where key is product name, value is the quantity sold, see example below

# 1. Find all transactions for a specific country (Germany)
# Example: es_transactions_in_germany = 7584
query_germany = {
    "query": {
        "match":{
            "Country": "Germany"
        }
    }
}

response_germany = es.count(index=index_name, body=query_germany)
es_transactions_in_germany = response_germany['count']

print(f"Jumlah transaksi di Jerman: {es_transactions_in_germany}")


Jumlah transaksi di Jerman: 9495


In [ ]:
# 2. Find all unique products in Germany
# Hints: Use set() to remove duplicate values
# Example: es_total_unique_products = 45345

query_unique_products = {
    "query": {
        "match": {
            "Country": "Germany"  # Field dan nilai yang dicari
        }
    },
    "aggs": {
        "unique_products": {
            "terms": {
                "field": "Description.keyword",
                "size": 100000
            }
        }
    },
    "size": 0  # Jangan ambil dokumen, hanya hasil agregasi
}

response_unique_products = es.search(index=index_name, body=query_unique_products)


buckets = response_unique_products['aggregations']['unique_products']['buckets']
unique_products = {bucket['key'] for bucket in buckets}  # Menggunakan set untuk menghapus duplikat

# Menghitung jumlah produk unik
es_total_unique_products = len(unique_products)

print(f"Jumlah produk unik di Jerman: {es_total_unique_products}")



Jumlah produk unik di Jerman: 1703


In [ ]:
# 3. Find the top 5 most purchased products in Germany
# Example: es_top_products = {'PRODUCT 1': 498, 'PRODUCT 2': 452, 'PRODUCT 3': 342, 'PRODUCT 4': 231, 'PRODUCT 5': 123}
query_top_products = {
    "query": {
        "match": {
            "Country": "Germany"
        }
    },
    "aggs": {
        "top_products": {
            "terms": {
                "field": "Description.keyword",
                "size": 5,
                "order": {
                    "total_quantity": "desc"
                }
            },
            "aggs": {
                "total_quantity": {
                    "value_count": {
                        "field": "Description.keyword"
                    }
                }
            }
        }
    },
    "size": 0
}

# Menjalankan query
response_top_products = es.search(index=index_name, body=query_top_products)

# Mengambil hasil agregasi
buckets = response_top_products['aggregations']['top_products']['buckets']
es_top_products = {bucket['key']: bucket['total_quantity']['value'] for bucket in buckets}

print(f"Top 5 produk yang paling banyak dibeli di Jerman: {es_top_products}")



Top 5 produk yang paling banyak dibeli di Jerman: {'POSTAGE': 383, 'ROUND SNACK BOXES SET OF4 WOODLAND ': 120, 'REGENCY CAKESTAND 3 TIER': 81, 'ROUND SNACK BOXES SET OF 4 FRUITS ': 78, 'PLASTERS IN TIN WOODLAND ANIMALS': 67}


In [ ]:
es_top_products

{'POSTAGE': 383,
 'ROUND SNACK BOXES SET OF4 WOODLAND ': 120,
 'REGENCY CAKESTAND 3 TIER': 81,
 'ROUND SNACK BOXES SET OF 4 FRUITS ': 78,
 'PLASTERS IN TIN WOODLAND ANIMALS': 67}

## Submission

Once you are satisfied with the performance of your model, then you run the code block below to submit your project.


In [29]:
# # Submit Method

# # Do not change the code below
# question_id = "01_sql_columns"
# submit(student_id, name, assignment_id, str(sql_columns), question_id, drive_link)
# question_id = "02_sql_num_rows"
# submit(student_id, name, assignment_id, str(sql_num_rows), question_id, drive_link)
# question_id = "03_sql_num_products"
# submit(student_id, name, assignment_id, str(sql_num_products), question_id, drive_link)
# question_id = "04_sql_num_customers"
# submit(student_id, name, assignment_id, str(sql_num_customers), question_id, drive_link)
# question_id = "05_sql_top_product_revenues"
# submit(student_id, name, assignment_id, str(sql_top_product_revenues), question_id, drive_link)
# question_id = "06_sql_top_countries"
# submit(student_id, name, assignment_id, str(sql_top_countries), question_id, drive_link)
question_id = "07_sql_top_selling_products_by_country_by_sales"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_sales), question_id, drive_link)
question_id = "08_sql_top_selling_products_by_country_by_quantity"
submit(student_id, name, assignment_id, str(sql_top_selling_products_by_country_by_quantity), question_id, drive_link)


'Assignment successfully submitted'

In [27]:
sql_top_selling_products_by_country_by_sales

{'United Kingdom': [('DOTCOM POSTAGE', 206245.48000000019),
  ('REGENCY CAKESTAND 3 TIER', 134405.93999999992),
  ('WHITE HANGING HEART T-LIGHT HOLDER', 93953.0699999987),
  ('PARTY BUNTING', 92501.73000000033),
  ('JUMBO BAG RED RETROSPOT', 84516.44000000166)],
 'Netherlands': [('RABBIT NIGHT LIGHT', 9568.48),
  ('ROUND SNACK BOXES SET OF4 WOODLAND ', 7991.400000000001),
  ('SPACEBOY LUNCH BOX ', 7485.5999999999985),
  ('DOLLY GIRL LUNCH BOX', 6828.599999999999),
  ('ROUND SNACK BOXES SET OF 4 FRUITS ', 4039.2000000000007)],
 'EIRE': [('REGENCY CAKESTAND 3 TIER', 7442.849999999997),
  ('CARRIAGE', 5175.0),
  ('JAM MAKING SET WITH JARS', 3089.0),
  ('3 TIER CAKE TIN RED AND CREAM', 3041.5499999999997),
  ('WHITE HANGING HEART T-LIGHT HOLDER', 2857.8)],
 'Germany': [('POSTAGE', 20821.0),
  ('REGENCY CAKESTAND 3 TIER', 8257.349999999997),
  ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3554.6999999999985),
  ('ROUND SNACK BOXES SET OF 4 FRUITS ', 1949.9500000000016),
  ('SPACEBOY LUNCH BOX ', 

In [28]:
sql_top_selling_products_by_country_by_quantity

{'United Kingdom': [('WORLD WAR 2 GLIDERS ASSTD DESIGNS', 48326),
  ('JUMBO BAG RED RETROSPOT', 43167),
  ('POPCORN HOLDER', 34365),
  ('ASSORTED COLOUR BIRD ORNAMENT', 33679),
  ('WHITE HANGING HEART T-LIGHT HOLDER', 33193)],
 'Netherlands': [('RABBIT NIGHT LIGHT', 4801),
  ('SPACEBOY LUNCH BOX ', 4528),
  ('DOLLY GIRL LUNCH BOX', 4132),
  ('PACK OF 72 RETROSPOT CAKE CASES', 4128),
  ('ROUND SNACK BOXES SET OF4 WOODLAND ', 3132)],
 'EIRE': [('PACK OF 72 RETROSPOT CAKE CASES', 1728),
  ('60 TEATIME FAIRY CAKE CASES', 1536),
  ('VINTAGE SNAP CARDS', 1492),
  ('ASSORTED INCENSE PACK', 1440),
  ('PACK OF 60 PINK PAISLEY CAKE CASES', 1296)],
 'Germany': [('ROUND SNACK BOXES SET OF4 WOODLAND ', 1218),
  ('ASSORTED COLOURS SILK FAN', 1164),
  ('POSTAGE', 1104),
  ('WOODLAND CHARLOTTE BAG', 1019),
  ('PACK OF 72 RETROSPOT CAKE CASES', 1002)],
 'France': [('RABBIT NIGHT LIGHT', 4023),
  ('MINI PAINT SET VINTAGE ', 2196),
  ('RED TOADSTOOL LED NIGHT LIGHT', 1314),
  ('SET/6 RED SPOTTY PAPER CUP

In [ ]:

# question_id = "09_es_columns"
# submit(student_id, name, assignment_id, str(es_columns), question_id, drive_link)
# question_id = "10_es_num_rows"
# submit(student_id, name, assignment_id, str(es_num_rows), question_id, drive_link)
# question_id = "11_es_total_unique_products"
# submit(student_id, name, assignment_id, str(es_total_unique_products), question_id, drive_link)
# question_id = "12_es_transactions_in_germany"
# submit(student_id, name, assignment_id, str(es_transactions_in_germany), question_id, drive_link)
# question_id = "13_es_top_products"
# submit(student_id, name, assignment_id, str(es_top_products), question_id, drive_link)

'Assignment successfully submitted'

## FIN